fft/rs/src/lib.rs

```diff
@@ -56,40 +56,130 @@ impl std::ops::Mul<f64> for Complex {
     }
 }
 
+fn bit_reverse_permute(arr: &mut [Complex]) {
+    let n = arr.len();
+    if n <= 1 {
+        return;
+    }
+    debug_assert!(n.is_power_of_two());
+
+    let bits: u32 = n.trailing_zeros();
+    for i in 0..n {
+        let j = i.reverse_bits() >> (usize::BITS - bits);
+        if j > i {
+            arr.swap(i, j as usize);
+        }
+    }
+}
+
 pub fn fft(arr: &mut [Complex]) {
-    fn _fft(arr: &mut [Complex]) {
-        let n = arr.len();
-        if n == 1 {
-            return;
+    let n = arr.len();
+    if n == 0 {
+        return;
+    }
+    debug_assert!(n.is_power_of_two());
+
+    // Bit-reversal permutation
+    bit_reverse_permute(arr);
+
+    // Iterative Cooley–Tukey, radix-2 DIT
+    let mut len = 2usize;
+    while len <= n {
+        let ang = -2.0 * PI / (len as f64);
+        let (s, c) = ang.sin_cos();
+        let wlen = Complex::new(c, s);
+        let half = len / 2;
+
+        let mut i = 0usize;
+        while i < n {
+            let mut w = Complex::new(1.0, 0.0);
+            let mut j = 0usize;
+            while j < half {
+                let u = arr[i + j];
+                let v = w * arr[i + j + half];
+                arr[i + j] = u + v;
+                arr[i + j + half] = u - v;
```

Comment:
Less bound checks, ~10ms improvement in my test
```suggestion
            let (_, arr) = arr.split_at_mut(i);
            let (left, right) = arr.split_at_mut(half);
            while j < half {
                let u_mut = &mut left[j];
                let v_mut = &mut right[j];
                let u = *u_mut;
                let v = w * *v_mut;
                *u_mut = u + v;
                *v_mut = u - v;
```

Reply:
I've added a Criterion benchmark, here's the baseline from my PR:

```
fft/2^18                time:   [6.1594 ms 6.1856 ms 6.2250 ms]
                        thrpt:  [42.111 Melem/s 42.380 Melem/s 42.560 Melem/s]
                 change:
                        time:   [+0.1919% +0.7281% +1.2549%] (p = 0.02 < 0.05)
                        thrpt:  [−1.2394% −0.7229% −0.1915%]
                        Change within noise threshold.
fft/2^20                time:   [27.379 ms 27.446 ms 27.521 ms]
                        thrpt:  [38.101 Melem/s 38.204 Melem/s 38.299 Melem/s]
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high mild
fft/2^22                time:   [121.97 ms 123.27 ms 124.65 ms]
                        thrpt:  [33.648 Melem/s 34.025 Melem/s 34.389 Melem/s]
```

<img width="1352" height="448" alt="image" src="https://github.com/user-attachments/assets/90603129-097e-4d76-9410-85bd60f304dd" />

And here's what it looks like with your patch:

```
fft/2^18                time:   [6.1979 ms 6.2292 ms 6.2459 ms]
                        thrpt:  [41.971 Melem/s 42.083 Melem/s 42.296 Melem/s]
                 change:
                        time:   [−0.3549% +0.1641% +0.6899%] (p = 0.57 > 0.05)
                        thrpt:  [−0.6851% −0.1638% +0.3562%]
                        No change in performance detected.
fft/2^20                time:   [27.615 ms 27.635 ms 27.665 ms]
                        thrpt:  [37.902 Melem/s 37.944 Melem/s 37.972 Melem/s]
                 change:
                        time:   [+0.0745% +0.4679% +0.8086%] (p = 0.03 < 0.05)
                        thrpt:  [−0.8021% −0.4657% −0.0745%]
                        Change within noise threshold.
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high mild
fft/2^22                time:   [121.92 ms 123.17 ms 124.67 ms]
                        thrpt:  [33.644 Melem/s 34.054 Melem/s 34.402 Melem/s]
                 change:
                        time:   [−1.5696% −0.0853% +1.5452%] (p = 0.92 > 0.05)
                        thrpt:  [−1.5217% +0.0853% +1.5946%]
                        No change in performance detected.
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high mild
```

<img width="1398" height="748" alt="image" src="https://github.com/user-attachments/assets/aaf1ae3e-feae-48c3-b4b6-33c2e364e754" />

Seems like it's slower or pure noise. Which is weird because your change made sense to me.

I rejiggered the benchmark into a head-to-head, here's what it looked like:

```
fft/fft/2^18            time:   [6.0983 ms 6.1031 ms 6.1140 ms]
                        thrpt:  [42.876 Melem/s 42.952 Melem/s 42.987 Melem/s]
fft/fft2/2^18           time:   [6.2194 ms 6.2544 ms 6.2849 ms]
                        thrpt:  [41.710 Melem/s 41.914 Melem/s 42.150 Melem/s]
Found 2 outliers among 10 measurements (20.00%)
  1 (10.00%) low mild
  1 (10.00%) high mild
fft/fft/2^20            time:   [27.235 ms 27.269 ms 27.309 ms]
                        thrpt:  [38.397 Melem/s 38.453 Melem/s 38.501 Melem/s]
Found 2 outliers among 10 measurements (20.00%)
  2 (20.00%) high mild
fft/fft2/2^20           time:   [27.490 ms 27.509 ms 27.529 ms]
                        thrpt:  [38.091 Melem/s 38.117 Melem/s 38.144 Melem/s]
Found 2 outliers among 10 measurements (20.00%)
  2 (20.00%) low mild
fft/fft/2^22            time:   [121.67 ms 123.67 ms 125.74 ms]
                        thrpt:  [33.358 Melem/s 33.916 Melem/s 34.472 Melem/s]
fft/fft2/2^22           time:   [121.90 ms 123.04 ms 124.28 ms]
                        thrpt:  [33.749 Melem/s 34.088 Melem/s 34.407 Melem/s]
```

<img width="1346" height="660" alt="image" src="https://github.com/user-attachments/assets/4dd9c043-b31e-4403-a7bb-0f1991b98875" />


<img width="960" height="540" alt="image" src="https://github.com/user-attachments/assets/c4f06a6b-f063-4f7f-877b-56b6f21f38ee" />

Seems like there's just a very small uplift at 2^22, but it's very slightly slower otherwise. Curious.

Reply:
Ah here we go, I found a way to make your version conclusively faster:

<img width="1392" height="1394" alt="image" src="https://github.com/user-attachments/assets/0782d1f3-e10a-4cbc-9cb9-42b5d45aeaed" />

I flipped `RUSTFLAGS="-C target-cpu=native"` on so the auto-vectorization would kick over. There you go.